pymc3/theanof.py

```diff
@@ -253,30 +253,33 @@ class GeneratorOp(Op):
     Generaror Op is designed for storing python generators
     inside theano graph. The main limitation is generator itself.
 
-    There are some important cases when generator becomes exhausted
-        - not endless generator is passed
-        - exception is raised while `generator.__next__` is performed
-            Note: it is dangerous in simple python generators, but ok in
-            custom class based generators with explicit state
-        - data type on each iteration should be the same
+    __call__ creates TensorVariable
+        It has 2 new methods (assigned externally)
+        - var.set_gen(gen) : sets new generator
+        - var.set_default(value) : sets new default value (None erases default value)
+
+    Parameters
+    ----------
+    gen : generator that implements __next__ (py3) or next (py2) method
+        and yields np.arrays with same types
+    default : np.array with the same type as generator produces
     """
     __props__ = ('generator',)
 
-    def __init__(self, gen):
+    def __init__(self, gen, default=None):
         super(GeneratorOp, self).__init__()
         if not isinstance(gen, DataGenerator):
             gen = DataGenerator(gen)
         self.generator = gen
         self.itypes = []
         self.otypes = [self.generator.tensortype]
-        self._nan = np.zeros_like(self.generator.test_value)
-        self._nan[...] = np.nan
+        self.set_default(default)
 
     def perform(self, node, inputs, output_storage, params=None):
-        try:
+        if self.default is not None:
+            output_storage[0][0] = next(self.generator, self.default)
```

Comment:
I can't comment on that next(p1, p2) syntax. I don't know it and a very quick google only talk about next(p1).

If what it does is to return default when the generator finished. This isn't clear in the doc. When this happen, should it print it that the generator finished?

For  NN, I don't know what would be a good default. The first minibatch? This would bias the training. What about raising StopIteration and making Theano catch that (and do some clean up when that happen and reraise it?)